explorer/explorer_ios.go

```diff
@@ -0,0 +1,255 @@
+
+//go:build ios
+// +build ios
+
+package explorer
+
+/*
+#cgo CFLAGS: -Werror -xobjective-c -fmodules -fobjc-arc
+
+#include <UIKit/UIKit.h>
+#include <stdint.h>
+
+@interface explorer_delegate:NSObject
+@end
+
+@interface explorer:NSObject<UIDocumentPickerDelegate>
+@property uint64_t mode;
+@property (strong) UIDocumentPickerViewController * picker;
+@property (strong) UIViewController * controller;
+@property uint32_t id;
+- (bool) show:(char*)text;
+- (bool) exportFile:(char*)name;
+- (bool) importFile:(char*)ext;
+@end
+
+static const uint64_t IMPORT_MODE = 1;
+static const uint64_t EXPORT_MODE = 2;
+
+static CFTypeRef createExplorer(uint64_t mode, CFTypeRef controllerRef, int32_t id) {
+	explorer * expl = [[explorer alloc] init];
+	expl.mode = mode;
+	expl.controller = (__bridge UIViewController *)controllerRef;
+	expl.id = id;
+	return (__bridge_retained CFTypeRef)expl;
+}
+
+static bool show(CFTypeRef expl, char * text) {
+	return [(__bridge explorer *)expl show:text];
+}
+
+static CFTypeRef fileWriteHandler(CFTypeRef u) {
+	NSError *err = nil;
+	NSFileHandle *handler = [NSFileHandle fileHandleForWritingToURL:(__bridge NSURL *)u error:&err];
+	if (err != nil) {
+		return 0;
+	}
+	return (__bridge_retained CFTypeRef)handler;
+}
+
+static CFTypeRef fileReadHandler(CFTypeRef u) {
+	NSError *err = nil;
+	NSFileHandle *handler = [NSFileHandle fileHandleForReadingFromURL:(__bridge NSURL *)u error:&err];
+	if (err != nil) {
+		return 0;
+	}
+	return (__bridge_retained CFTypeRef)handler;
+}
+
+static void closeFile(CFTypeRef handler, CFTypeRef u) {
+	[(__bridge NSURL *)u stopAccessingSecurityScopedResource];
+	[(__bridge NSFileHandle *)handler closeFile];
+}
+
+static void fileWrite(CFTypeRef handler, uint8_t* b, uint64_t len) {
+	NSData *data = [NSData dataWithBytes:b length:len];
+
+	NSFileHandle * h = (__bridge NSFileHandle *)handler;
+	[h writeData:data];
+}
+
+static uint64_t fileRead(CFTypeRef handler, uint8_t* b, uint64_t len) {
+	if (@available(iOS 14, *)) {
+		NSError *err = nil;
+
+		NSFileHandle * h = (__bridge NSFileHandle *)handler;
+		NSData *data = [h readDataUpToLength:len error:&err];
+		[data getBytes:b length:data.length];
+
+		return data.length;
+	}
+	return 0;
+}
+*/
+import "C"
+import (
+	"gioui.org/app"
+	"gioui.org/io/event"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"unsafe"
+)
+
+type explorer struct {
+	window     *app.Window
+	mutex      sync.Mutex
+	controller C.CFTypeRef
+	importFile C.CFTypeRef
+	exportFile C.CFTypeRef
+	result     chan result
+}
+
+func newExplorer(w *app.Window) *explorer {
+	return &explorer{window: w, result: make(chan result)}
+}
+
+func (e *Explorer) listenEvents(evt event.Event) {
+	switch evt := evt.(type) {
+	case app.ViewEvent:
+		e.controller = C.CFTypeRef(evt.ViewController)
+
+		e.explorer.exportFile = C.createExplorer(C.EXPORT_MODE, e.controller, C.int32_t(e.id))
+		e.explorer.importFile = C.createExplorer(C.IMPORT_MODE, e.controller, C.int32_t(e.id))
```

Comment:
I don't think I understand why we create two explorers (one in each mode) rather than have a single explorer with the mode either encoded as a parameter to the methods or in the name of the methods. Can you explain why this is necessary?

Reply:
That was to "simplify the callback" and the "show" function. Since in both cases it calls `documentPicker` or `documentPickerWasCancelled`. In that case, the `documentPicker` for  `e.explorer.exportFile` will always call `exportCallback`.

However, that is not necessary, since it can change based on `show` function. So, if call `show(C.EXPORT_MODE)` it sets `@property uint64_t mode;` to `C.EXPORT_MODE` and then the callback is `exportCallback`. It will change if `show(C.IMPORT_MODE)` is called.

Reply:
Let's change the mode based on the show function then.

Reply:
I have mitigated it on my latest commit (`explorer: [ios] move functions to .m file`), since the `exportFile` and `importFile` are directly exposed.